agents/ansari_workflow.py

```diff
@@ -0,0 +1,193 @@
+import hashlib
+import logging
+import os
+import sys
+from datetime import date
+
+import litellm
+
+from tools.search_hadith import SearchHadith
+from tools.search_vectara import SearchVectara
+from tools.search_quran import SearchQuran
+from util.prompt_mgr import PromptMgr
+
+logger = logging.getLogger(__name__ + ".AnsariWorkflow")
+
+if not sys.argv[0].endswith("main_api.py"):
+    logging_mode = logging.DEBUG
+else:
+    logging_mode = logging.INFO
+
+logger.setLevel(logging_mode)
+
+# # Uncomment below when logging above doesn't output to std, and you want to see the logs in the console
+# console_handler = logging.StreamHandler()
+# console_handler.setLevel(logging_mode)
+# logger.addHandler(console_handler)
+
+
+class AnsariWorkflow:
+    """
+    AnsariWorkflow manages the execution of modular workflow steps for processing user queries.
+
+    This class provides a flexible framework for generating queries, performing searches,
+    and generating answers based on the results. It supports customizable workflow steps
+    that can be combined in various ways to handle different types of user interactions.
+
+    Attributes:
+        tool_name_to_instance (dict): Mapping of tool names to their respective instances.
+        model (str): The name of the language model to use.
```

Comment:
Todo (for future): Support different models in different parts of the workflow.

Reply:
One thing I also don't understand: why doesn't this replace agents/ansari.py?

Reply:
This implementation is based on our previous WhatsApp discussion. We identified two distinct use cases:

1. Letting the LLM dynamically decide whether and which tools to use
2. Enforcing a specific sequence of tool usage

For some scenarios (like the quran.com integration), we want to guarantee a specific workflow - first execute a search, then formulate an answer based on the search results. Therefore:
- `Ansari` will handle cases where the LLM makes tool-usage decisions
- `AnsariWorkflo` will execute predefined sequences of steps